Checks on itlwm

Aldo97 opened the pull request:
Hi,
I modified the AppDelegate showing an alert when HeliPort is launched without itlwm installed. Pressing "Quit" nothing happens and the app closes, pressing "Continue" the app is launched and the app stores a bool named "itlwm". Now the app doesn't alert the user until the user launches HeliPort with itlwm running. In this case HeliPort change the value of "itlwm" and the next time HeliPort is launched without itlwm running the alert would be back.
I modified also the alert for the API mismatch. In this case the behaviour is similar to above but the button "Open OpenIntelWireless on GitHub" would be shown only the first time the mismatch is detected. If the user launches HeliPort with the correct version of itlwm, also this behaviour is resetted.
Worth noting that unfortunately this pull request includes also the changes of #86, I'm new at GitHub and I built the new branch "MIsmatch" after the commits of the precedent pull request.
If the request is granted it is needed to upgrade the translations.

Aldo97 commented:
It could be possible also to add a checkbox to both the alerts in order to make the user decide if he wants to not be alerted again. Obviously the "reset" behaviour should be kept intact.
For itlwm mismatch I think, instead of the "Visit OpenIntelWireless on Github" button a "Download/Install itlwm" button in which the app automatically downloads itlwm from the github releases. 
It could be made also a script to automatically install the new update in clover, OC or /L/E.
If you like the idea I can work on it.

williambj1 commented:
Hi @Aldo97 

Thank you for your contribution, I have some questions and comments about this PR.

> Now the app doesn't alert the user until the user launches HeliPort with itlwm running.

Could you please explain how would this benefit regular users? Based on my understanding, this only benefits developers. I'm planning on adding some Macros to control the behavior of Debug and Release builds. Please correct me if I'm wrong.

> "Open OpenIntelWireless on GitHub" would be shown only the first time the mismatch is detected.
>
> It could be possible also to add a checkbox to both the alerts in order to make the user decide if he wants to not be alerted again. Obviously the "reset" behaviour should be kept intact.

Actually I don't want to leave any tolerance for regular users, people should really use the matched version, and it's fairly easy to get rid of this alert; all they need to do is download the correct version.

> For itlwm mismatch I think, instead of the "Visit OpenIntelWireless on Github" button a "Download/Install itlwm" button in which the app automatically downloads itlwm from the github releases.
It could be made also a script to automatically install the new update in clover, OC or /L/E.

Sounds like a good idea, but I really doubt the actual benefit of implementing it.
1. In my opinion, users who play hacks must understand how to deal with their EFIs.
2. It's quite hard for us to deal with EFI partitions since they could exist in all storage devices, including removable ones. GPT even allows users to have multiple EFI partitions on a single hard drive. What's impossible for us to get the loading path of itlwm meaning that we need to search all possible paths to find the kext.

I think it's better to leave it to the users, so I decided to guide them to GitHub.

Aldo97 commented:
> Could you please explain how would this benefit regular users? Based on my understanding, this only benefits developers. I'm planning on adding some Macros to control the behavior of Debug and Release builds. Please correct me if I'm wrong.
>
Indeed a regular user has no benefit from this. It could also be included in the ifs that I have added a condition on whether HeliPort is in debug or release. If the things are not changed it should be solved by a simple "if DEBUG"

> Actually I don't want to leave any tolerance for regular users, people should really use the matched version, and it's fairly easy to get rid of this alert; all they need to do is download the correct version.
>
Yes, I think it is a sensible idea.

> 2. It's quite hard for us to deal with EFI partitions since they could exist in all storage devices, including removable ones. GPT even allows users to have multiple EFI partitions on a single hard drive. What's more, it's impossible for us to get the loading path of itlwm meaning that we need to search all possible paths to find the kext.
> 
The script can simply search in /L/E and /S/L/E, if itlwm is not found it searches all the EFI on the PC and if more than 1 EFI is found (with the same itlwm version running) asks the user to chose the correct EFI. It's not ideal but it's not hard to implement and for the most of the users it would work without the need to ask which EFI partition to use.
In case itlwm is running but it is not installed, the script would only download the release (or the debug version if the running HeliPort is in debug). The same happens if the script can't find an itlwm file that matches the one running (the removable devices is removed before the launch of HeliPort).
Obviously the first thing that would be asked is if the user want only to download or also install itlwm, in this case any issue with user that uses removable EFI is solved as they can be alerted.
If this features is implemented it could be possible to add, at first boot and only for release, a similar script that asks(or automatically decide what to do checking if the kexts are installed or injected via OC/Clover) the user where to install itlwm if it detects that itlwm is not running

ErrorErrorError commented:
> It's not ideal but it's not hard to implement

I think it would much be easier for the user to just download the latest version from there, and then let the user install it to their EFI, rather than having a whole program dedicate itself in installing it to the "assumed" correct EFI folder. Sound's like there is much more steps involved trying to find an EFI folder and dealing with permissions and more error prone.

Aldo97 commented:
> > It's not ideal but it's not hard to implement
> 
> I think it would much be easier for the user to just download the latest version from there, and then let the user install it to their EFI, rather than having a whole program dedicate itself in installing it to the "assumed" correct EFI folder. Sound's like there is much more steps involved trying to find an EFI folder and dealing with permissions and more error prone.

A regular user would not have ambiguity on the EFI as normally should have only one EFI with clover/OC even if it has more disks on the system. In case of multiple clover/OC installations I think that it is still simpler to alert the user with all the EFI available.
The only critical scenarios, if I'm not missing something, are when the user boots from a removable device(that is removed after the boot) and has itlwm also in a EFI on the disk, or if the user unload the kext and then reload it. In the first case an alert would be enough to solve the problem assuming that the user knows what's doing. In the second case the problem would be only if the user unloads and loads the same version of itlwm, something that is not "normal". Even if the same itlwm version is present in an unused clover/OC folder and loaded in the OS via terminal the alert I talked above should be enough to remove the ambiguity.

williambj1 commented:
To be honest, users should be capable to handle the kext update themselves. I'm worried that the time and effort of our work to make it perfect does not worth the final result.

Aldo97 commented:
With the commit above I have implemented #if debug. Now if the app is a release and there is a mismatch/missing itlwm the app closes after pressing "Visit OpenIntelWireless". 
In case of debug build, pressing on "Visit OpenIntelWIreless" HeliPort starts normally and opens also the GitHub page. At next app launch the button becames "Continue to HeliPort" without opening the webpage. This behaviour is resetted when the itlwm build is running before a the launch of HeliPort and/or when the itlwm version is upgraded.
I added also a reset of the variables when the user switch from debug to release.


> To be honest, users should be capable to handle the kext update themselves. I'm worried that the time and effort of our work to make it perfect does not worth the final result.
As you wish, thanks.

williambj1 commented:
@Aldo97 In my opinion, I would prefer this behavior. It's much simpler and it doesn't require User Defaults

- Debug
   - itlwm not loaded
      Don't show any alerts and make it friendly for developers.
   - API version mismatch
      Don't show any alerts and make it friendly for devs.
- Release
   - itlwm not loaded
      Show an alert with only one button, tell the user that itlwm is not loaded and continue launching the app
   - API version mismatch
      Show an alert every time the app launches with two options, one guides the user to OpenIntelWireless and continue launching the app, another quits the app.
      It's fair to show open an **annoying webpage everytime** then let the users to continue using the app. Let's say if someone updated itlwm but forgot to update HeliPort, the only solution for him to connect to the Internet is using HeliPort (most of the time APIs don't change that much so he could still connect). Then he could download the correct version from Github or perform an in-app-update

Aldo97 commented:
@williambj1
I see what you say and I'm ok in all but one thing.
Even if the developers should be able to know that a new versions of itlwm is out is still a good thing, in my opinion, to alert that the APIs mismatch.
One developer could be working only on HeliPort, with no interest on itlwm.

Goshin commented:
To me, a simple warning like "itlwm not detected" would be quite enough. No need for a "ignore" checkbox or a debug flag to mute the alert, as the developers can easily bypass it by commenting out the lines, and users do need a good indication of whether itlwm is running properly everytime they launch the app. Just keep it simple.

Aldo97 commented:
Now it should work (didn't explored all the cases). 
It is needed also to add the new line in the translation files.

williambj1 commented:
I would add the translation files after this PR is merged. Thanks.

Aldo97 commented:
@williambj1
I modified the strings and added the Italian translation as I "support" the Italian translation.
If the strings are ok in English I have no problem to add them in all the other translation files.
I tried all the situations and it's working as expected. Translations and strings apart it is ready to merge.

williambj1 commented on HeliPort/AppDelegate.swift:
```diff
@@ -38,38 +38,59 @@ class AppDelegate: NSObject, NSApplicationDelegate, NSMenuDelegate {
+                verAlert.addButton(withTitle: NSLocalizedString("Quit HeliPort", comment: "")).keyEquivalent = "\r"
+            } else {
+                verAlert.messageText = NSLocalizedString("itlwm is not running", comment: "")
+                    verAlert.informativeText = NSLocalizedString("Install or load itlwm v.", comment: "") + String(IOCTL_VERSION)
```
Please remove this line, `IOCTL_VERSION` is Heliport's API but not itlwm's `MARKETING_VERSION`

Aldo97 replied:
Done

Goshin commented:
Resolve conflicts and refactor. Please add translation in another PR later.

Aldo97 commented:
@Goshin I think it's ready to merge

williambj1 commented:
Merged, thanks.
